storage/client_raft_test.go

```diff
@@ -1165,3 +1165,67 @@ func TestStoreRangeRemoveDead(t *testing.T) {
 	}
 	ticker.Stop()
 }
+
+// TestStoreRangeDownReplicate verifies that the replication queue will notice
+// over-replicated ranges and remove replicas from them.
+func TestStoreRangeRebalance(t *testing.T) {
+	defer leaktest.AfterTest(t)
+
+	// Start multiTestContext with replica rebalancing enabled.
+	mtc := &multiTestContext{
+		storeContext: &storage.StoreContext{},
+	}
+	*mtc.storeContext = storage.TestStoreContext
+	mtc.storeContext.RebalancingOptions = storage.RebalancingOptions{
+		AllowRebalance: true,
+		Deterministic:  true,
+	}
+
+	// Four stores.
+	mtc.Start(t, 4)
+	defer mtc.Stop()
+
+	// Replicate the first range to the first three stores.
+	store0 := mtc.stores[0]
+	replica := store0.LookupReplica(proto.KeyMin, nil)
+	desc := replica.Desc()
+	mtc.replicateRange(desc.RangeID, 0, 1, 2)
+
+	// Initialize the gossip network with fake capacity data.
+	storeDescs := make([]*proto.StoreDescriptor, 0, len(mtc.stores))
+	for _, s := range mtc.stores {
+		desc, err := s.Descriptor()
+		if err != nil {
+			t.Fatal(err)
+		}
+		desc.Capacity.Capacity = 1024 * 1024
+		desc.Capacity.Available = 1024 * 1024
+		// Make sure store[1] is chosen as removal target.
+		if desc.StoreID == mtc.stores[1].StoreID() {
+			desc.Capacity.Available = 0
+		}
+		storeDescs = append(storeDescs, desc)
+	}
+	sg := gossiputil.NewStoreGossiper(mtc.gossip)
+	sg.GossipStores(storeDescs, t)
+
+	// Within a few seconds, a replica should be added to the fourth store if we
+	// keep forcing replication scans.
```

Comment:
Why does it need multiple scans? Shouldn't a single replication scan be enough to start the rebalance process since you're in deterministic mode? You'd still need an `IsTrueWithin` to check that the range descriptor is asynchronously updated but I don't see why you'd need to repeat the scan.